### PR TITLE
[Feature - AOIs] AOIs profile form

### PR DESCRIPTION
**Pull request by blayhem**
## Overview

WIP Profile modal to update profile data from myGFW. Introduces new fields such as firstName, lastName, topics interested in, aoiLocation, etc.

image

### CODE REVIEW

File: app/javascript/components/forms/components/field-wrapper/styles.scss [hunk 1]
```diff
@@ -2,18 +2,10 @@
 
 .c-form-field {
   font-size: rem(14px);
-  margin-bottom: rem(25px);
+  margin: rem(12px) 0;
```
Comment (edbrett) [inline app/javascript/components/forms/components/field-wrapper/styles.scss]: Why this change?

File: app/javascript/components/forms/components/field-wrapper/styles.scss [hunk 2]
```diff
@@ -57,6 +49,14 @@
     transition: all 0.1s ease-in-out;
   }
 
+  details {
```
Comment (edbrett) [inline app/javascript/components/forms/components/field-wrapper/styles.scss]: Error with the linter?

File: app/javascript/components/forms/profile/component.jsx [hunk 3]
```diff
@@ -78,47 +81,74 @@ class ProfileForm extends PureComponent {
...
+                        placeholder="Select a sector"
                       />
+                      {values.sector &&
+                        sectors[values.sector] && (
+                        <Radio
+                          name="subsector"
+                          label="Role"
+                          options={sectors[values.sector].map(s => ({
+                            label: s,
+                            value: s.replace(/( )+|(\/)+/g, '_')
```
Comment (edbrett) [inline app/javascript/components/forms/profile/component.jsx]: Not sure this is totally necessary for this native select.

Comment (edbrett) [reply app/javascript/components/forms/profile/component.jsx]: In fact maybe we should refactor it to only accept arrays or string and scrap the objects all together.

File: app/javascript/components/forms/components/radio/component.jsx [hunk 4]
```diff
@@ -54,6 +64,17 @@ class Radio extends PureComponent {
                       <label className="radio-label" htmlFor={id}>
                         <span />
                         {option.label}
+                        {otherRegex.test(option.label) &&
+                          selectedOption &&
+                          otherRegex.test(selectedOption) && (
+                          <Field
```
Comment (edbrett) [inline app/javascript/components/forms/components/radio/component.jsx]: Im not a big fan of including this inside this component. It feels like a very specific thing to the profile form. Surely just a conditional field element below in the form is simple enough? Also the styles of the input you have included are not consistent with any other input field. It should be below and the same height etc.

Screenshot 2020-02-04 at 16 15 23

File: app/javascript/components/forms/components/field-wrapper/styles.scss (hunk 1 above)
Comment (blayhem) [reply app/javascript/components/forms/components/field-wrapper/styles.scss]: To add margin top to the form fields, so that they have separation on both sides (for example, to the h4 titles).

File: app/javascript/components/forms/components/field-wrapper/styles.scss (hunk 2 above)
Comment (blayhem) [reply app/javascript/components/forms/components/field-wrapper/styles.scss]: yes  🤷‍♂   same functionality

File: app/javascript/components/forms/profile/component.jsx (hunk 3 above)
Comment (blayhem) [reply app/javascript/components/forms/profile/component.jsx]: yeah I agree we should do this in the component, as it's repeated a couple of times, but I wasn't sure if we were gonna keep that after the redesign

Comment (edbrett) [reply app/javascript/components/forms/profile/component.jsx]: Lets update it then. Redesign?

File: app/javascript/components/forms/components/radio/component.jsx (hunk 4 above)
Comment (blayhem) [reply app/javascript/components/forms/components/radio/component.jsx]: it's half the height on purpose, otherwise it moves the "other" option too far from the others. It was a decision not very consistent with the UI kit, but in order to respect the parent component first. 

It could be a conditional field, yes, passed as a prop. I did this for any fields having "other" and "write" keywords, as it sounds like a common case for radio buttons, but it could be an extra field in the option object, agree (less specific that way).

Comment (edbrett) [reply app/javascript/components/forms/components/radio/component.jsx]: In this case we need to ask the designers.

File: app/javascript/components/forms/components/field-wrapper/styles.scss (hunk 1 above)
Comment (edbrett) [reply app/javascript/components/forms/components/field-wrapper/styles.scss]: Ok. For me the h4 titles are something custom to your individual form so should manage their layout and styles personally. Our form fields should be put components that always do the same things and make forms consistent. Also we don't have a design for the form with the new fields so the h4 titles do not make sense just yet. Please discuss with a designer.

File: app/javascript/components/forms/profile/component.jsx [hunk 5]
```diff
@@ -61,15 +63,16 @@ class ProfileForm extends PureComponent {
...
+                    <div className="column small-12 medium-12">
                       <h1>Your profile information</h1>
                       <h3>
                         Help us help you! Tell us who you are and how you use
                         Global Forest Watch so we can better meet your needs.
                       </h3>
                     </div>
-                    <div className="column small-12 medium-6">
-                      <Input name="fullName" label="name" required />
+                    <div className="column small-12 medium-12">
```
Comment (edbrett) [inline app/javascript/components/forms/profile/component.jsx]: Why medium-12 if we already have the small-12? They both do the same thing?

File: app/javascript/components/forms/profile/component.jsx [hunk 6]
```diff
@@ -61,15 +63,16 @@ class ProfileForm extends PureComponent {
                   </div>
                 ) : (
                   <Fragment>
-                    <div className="column small-12">
+                    <div className="column small-12 medium-12">
```
Comment (edbrett) [inline app/javascript/components/forms/profile/component.jsx]: Same here

File: app/javascript/components/forms/profile/styles.scss [hunk 7]
```diff
@@ -9,7 +9,7 @@
 
   h3 {
     font-size: rem(16px);
-    margin-bottom: rem(40px);
+    margin-bottom: rem(28px);
```
Comment (edbrett) [inline app/javascript/components/forms/profile/styles.scss]: Reason for change?

Comment (blayhem) [reply app/javascript/components/forms/profile/styles.scss]: The top margin I added to the fields. Rolling back.

**Comment by edbrett** [comment]
One thing I noticed when reviewing the staging app, let's set a max width for input fields. I haven't set this  inside the form components as in some cases we use a grid and that would break the flow, so I have been doing it on a form by form basis. Here is how I did it for contact us: https://github.com/Vizzuality/gfw/blob/develop/app/javascript/components/forms/contact/styles.scss#L18

**Comment by aabdaab** [comment]
checking

File: app/javascript/components/forms/profile/component.jsx [hunk 8]
```diff
@@ -77,55 +86,108 @@ class ProfileForm extends PureComponent {
...
-                        options={sectors.map(s => ({ label: s, value: s }))}
-                        placeholder="Select a language"
-                        required
+                        options={Object.keys(sectors).map(s => ({
+                          label: s,
+                          value: s
+                        }))}
+                        placeholder="Select a sector"
+                        required={
+                          source === 'AreaOfInterestModal' ? true : undefined
```
Comment (aabdaab) [inline app/javascript/components/forms/profile/component.jsx:98]: never set a variable to `undefined` by yourself - this is reserved to javascript engine - because when debugging you cannot tell if the value is undefined because you never defined value and js engine set it to undefined (all vars in js are initially set to undefined) or if you did somewhere. it can make debugging very confusing :P

Comment (aabdaab) [reply app/javascript/components/forms/profile/component.jsx:98]: especially here, this can be solved with `required={source === 'AreaOfInterestModal'}`, if the source is something else, it's gonna evaluate to false

File: app/javascript/components/modals/profile/component.jsx [hunk 9]
```diff
@@ -18,16 +19,16 @@ class ProfileModal extends PureComponent {
   };
 
   render() {
-    const { open } = this.props;
+    const { open, filledProfile } = this.props;
 
     return (
       <Modal
-        isOpen={!!open}
+        isOpen={!!open || !filledProfile}
```
Comment (aabdaab) [inline app/javascript/components/modals/profile/component.jsx:26]: There's one case that isn't covered and it's the one with new users creating an account:
1. Users who create a MyGFW profile (not as part of saving an area) will be taken to the MyGFW profile modal window ✅ - that's working
2. Save button will be disabled until required fields complete, but the user can exit out of form without completing required fields - ❌ - that's not working, the user cannot exit the form.

We can handle that _without_ introducing a new variable `dismissed` or anything, it's better to rely on a single source of truth. We should rely solely on `open` variable and on the component _mount_ we can set the modal open if the profile isn't filled.
So:
`<Modal isOpen={open} ... >`
and in addition add componentDidMount lifecycle method
```
componentDidMount() {
    const { filledProfile, setProfileModalOpen } = this.props;
    if (!filledProfile) {
      setProfileModalOpen(true);
    }
  }
```

File: app/javascript/components/forms/profile/index.js [hunk 10]
```diff
@@ -3,13 +3,37 @@ import { connect } from 'react-redux';
...
 import * as actions from './actions';
 
+const splitFirstName = fullName => {
+  const arr = fullName.split(' ');
+  return arr.slice(0, arr.length / 2).join(' ');
+};
+
+const splitLastName = fullName => {
+  const arr = fullName.split(' ');
+  return arr.slice(arr.length / 2, arr.length).join(' ');
```
Comment (aabdaab) [inline app/javascript/components/forms/profile/index.js:13]: or just `arr.pop()`

File: app/javascript/components/modals/area-of-interest/component.jsx [hunk 11]
```diff
@@ -29,7 +30,9 @@ class AreaOfInterestModal extends PureComponent {
   };
 
   render() {
-    const { open, loading, loggedIn, canDelete, viewAfterSave } = this.props;
+    const { open, loading, userData, canDelete, viewAfterSave } = this.props;
+    const { email, fullName, lastName, loggedIn } = userData || {};
+    const isProfileFormFilled = !!email && (!!fullName || !!lastName);
```
Comment (aabdaab) [inline app/javascript/components/modals/area-of-interest/component.jsx:35]: you don't need to do this; `email && (fullName || lastName)` will work just fine

File: app/javascript/components/forms/profile/component.jsx (hunk 8 above)
Comment (blayhem) [reply app/javascript/components/forms/profile/component.jsx:98]: So in this case required could be false as you said. I'm using undefined because it's the same thing as not passing the prop 'required'. Ugly but that's how it is. I'm doing this to let the library manage if it's required or not, but for now false would be more than enough, you're right 👍

File: app/javascript/components/modals/profile/component.jsx (hunk 9 above)
Comment (blayhem) [reply app/javascript/components/modals/profile/component.jsx:26]: yes, we talked about this, and it's pending. I like your approach, better than using the state 🙂 thanks!
